Create open sourcing post (#73)
hackclub/hcb-engr: content/posts/open-source/post.mdx

> # bank.engineering post

Reviewer: ```suggestion
# The journey to open sourcing a $50 million neobank
```

### Security discussions

We had a lot of back and forward conversations about security during this process. We’ve implemented a couple of safeguards during this process but the general consensus walking away from those conversations was that we were risking very little by open sourcing and long-term, our codebase would be more secure. 

### Documentation

Lastly, we wanted to make sure that our codebase would be accessible to new developers. We worked on making the setup process smoother w/ Codespaces, wrote guides to the key parts of HCB (for example, [this guide on the transaction mapping engine](https://github.com/hackclub/hcb/blob/main/dev-docs/guides/transaction_imports.md)), and I gave a talk at SF Ruby with a brief overview of our codebase:

<iframe height="400" src="https://www.youtube.com/embed/CBxilReUkJ0?si=GgcQ9GDoqSl1avoK;start=3005" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" style={{borderRadius: "8px", width: "100%"}} allowfullscreen=""></iframe>

> There’s still more work to do! Reach out if you wish a part of HCB was documented better (hcb-engr@hackclub.com). And, if you’re a teenager interested in helping out, join us in the `#hcb-dev` channel on the [Hack Club Slack](https://hackclub.com/slack).

Reviewer: Add referer/UTM param to slack link

import post from '@/components/Post';

export default post({
    title: "💻 How we open sourced HCB",
    slug: "open-source",
    category: "new",
    authors: ["sam"],
    date: new Date("2025-03-29T20:00:00.200Z"),
});

# bank.engineering post

> [Last night we open-sourced HCB](https://hackclub.com/fiscal-sponsorship/open-source/), taking a seven-year old codebase from private to public has been long journey. Check it out and star it here: [github.com/hackclub/hcb](https://github.com/hackclub/hcb).

Reviewer: reader question: What does it mean to open source?

Here’s how we did it.

### Scrubbing

> The biggest part of the process was removing all the customer date / personally identifiable data from our repository. Over time, with bug reports etc., we’d posted multiple things that were highly sensitive. The only problem - we had nearly 5,000 issues & PRs which could potentially contain this information. We attempted to automate this process with Regex but it was difficult to cover edge case. Our solution was to hire a team of “scrubbers” who would go through the repository and flag any sensitive content. We assigned every scrubber a block of hundred issues / PRs; once finished they could claim a new block of issues and PRs. Within fifteen days - the entire repository was scrubbed!

Reviewer: ```suggestion
HCB began as a private repository with the intention of open sourcing one day. With it being private, the repository collected amounts of customer data / personally identifiable data (PII). Over time, with bug reports, etc., we posted highly sensitive information across GitHub Issues, Pull Requests, and commits. 

The only problem - we had nearly 5,000 issues & PRs which could potentially contain this information. We attempted to automate this process with Regex, but it was difficult to cover edge cases such as PII within images. Our solution was to hire a team of “scrubbers” who would go through the repository and flag any sensitive content. We assigned every scrubber a block of hundred issues / PRs; once finished they could claim a new block of issues and PRs. Within fifteen days - the entire repository's Issues and Pull Requests were scrubbed of customer data and PII.
```

![](Screenshot%202025-03-29%20at%2010.46.46%E2%80%AFAM.png)

> The scrubbers ended up identifying 170 instances of sensitive information in the repository. I then took that list and reached out to GitHub support (thank you Fred!) to remove the images from GitHub servers and the issues / edit history from their database.

Reviewer: ```suggestion
The scrubbers ended up identifying 170 instances of sensitive information in the repository. I then took that list, updated the GitHub comments to remove those instances, and reached out to GitHub support (thank you Fred!) to remove the images from GitHub servers and the edit history from their database.
```

### Credentials w/ Doppler

> Our next concern was our credentials. Don’t worry, we hadn’t been storing our `.env` in the repository. We were using [Rails’ encrypted credentials system](https://guides.rubyonrails.org/security.html#environmental-security). It’s decent, however, we were concerned that if our key leaked, everything would be leaked. With that in mind, we transitioned to storing our credentials with [Doppler](https://www.doppler.com/) and [built a custom module](https://github.com/hackclub/hcb/blob/main/app/lib/credentials.rb) to load the credentials into the environment. Some benefits of using Doppler: more flexibility with environments, audit logs, and convenient changes. The module also supported overriding the variables in Doppler in an ignored `.env`; this is also the system for developers without Doppler access.

Reviewer: ```suggestion
Our next concern was our credentials. Don’t worry, we hadn’t been storing our `.env` in the repository. We were using [Rails’ encrypted credentials system](https://guides.rubyonrails.org/security.html#environmental-security). It’s decent, however, we were concerned that if our key leaked, everything would be leaked with no option for remediation. With that in mind, we transitioned to storing our credentials with [Doppler](https://www.doppler.com/) and [built a custom module](https://github.com/hackclub/hcb/blob/main/app/lib/credentials.rb) to load the credentials into the Rails processes' environment. Some benefits of using Doppler: more flexibility with environments, audit logs, and convenient changes. The module also supported overriding Doppler variables for specific environments using an ignored `.env`. For developers without Doppler access, this is also the system for setting environment variables.
```

>more flexibility with environments

>convenient changes

elaborate more here

import post from '@/components/Post';

export default post({

>     title: "💻 How we open sourced HCB",

Author: ```suggestion
    title: "💻 The journey to open sourcing a $50 million neobank",
```
